File: java/protocol/src/test/java/io/github/yannisyoussef/qe/report/protocol/CorpusTest.java
package io.github.yannisyoussef.qe.report.protocol;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.DynamicTest.dynamicTest;

import com.fasterxml.jackson.databind.JsonNode;
import io.github.yannisyoussef.qe.report.protocol.testing.Corpus;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

/** The fixture corpus is the executable specification; every entry of the manifest is a test. */
class CorpusTest {
  private static final com.fasterxml.jackson.databind.json.JsonMapper TYPE_MAPPER =
      com.fasterxml.jackson.databind.json.JsonMapper.builder().build();
  private static final Set<String> CODEC_REASONS =
      Set.of(
          "MALFORMED_JSON",
          "SCHEMA_INVALID",
          "UNSUPPORTED_PROTOCOL_VERSION",
          "UNSUPPORTED_EVENT_TYPE");

  @TestFactory
  List<DynamicTest> validEvents() {
    List<DynamicTest> tests = new ArrayList<>();
    for (JsonNode entry : Corpus.manifest().get("events").get("valid")) {
      String file = entry.get("file").asText();
      boolean exact = "exact".equals(entry.get("roundTrip").asText());
      tests.add(
          dynamicTest(
              file,
              () -> {
                String text = Corpus.readText(Corpus.fixtures().resolve(file));
                Event event = ProtocolJson.read(text);
                if (event.payload() instanceof UnknownPayload) {
                  assertTrue(event.isIgnorable(), "unknown type must be ignorable");
                } else {
                  assertTrue(
                      SchemaValidation.isValid(text),
                      () -> SchemaValidation.describe(SchemaValidation.errors(text)));
                }
                String written = ProtocolJson.write(event);
                if (!(event.payload() instanceof UnknownPayload)) {
                  assertTrue(
                      SchemaValidation.isValid(written),
                      () -> SchemaValidation.describe(SchemaValidation.errors(written)));
                }
                if (exact) {
                  assertEquals(Corpus.canonical(text), Corpus.canonical(written));
                }
                String again = ProtocolJson.write(ProtocolJson.read(written));
                assertEquals(Corpus.canonical(written), Corpus.canonical(again), "idempotent");
              }));
    }
    return tests;
  }

  @TestFactory
  List<DynamicTest> invalidEvents() {
    List<DynamicTest> tests = new ArrayList<>();
    for (JsonNode entry : Corpus.manifest().get("events").get("invalid")) {
      String file = entry.get("file").asText();
      String reason = entry.get("reason").asText();
      JsonNode pointerNode = entry.get("pointer");
      String pointer = pointerNode.isNull() ? null : pointerNode.asText();
      boolean reject = "reject".equals(entry.get("codec").asText());
      tests.add(
          dynamicTest(
              file + " -> " + reason,
              () -> {
                String text = Corpus.readText(Corpus.fixtures().resolve(file));
                if (reason.equals("SCHEMA_INVALID")) {
                  assertFalse(SchemaValidation.errors(text).isEmpty(), "schema must reject");
                  String type = TYPE_MAPPER.readTree(text).path("eventType").asText();
                  var typed =
                      EventTypes.isKnown(type)
                          ? SchemaValidation.errorsForType(text, type)
                          : SchemaValidation.errors(text);
                  assertFalse(typed.isEmpty(), "typed schema must reject");
                  if (pointer != null) {
                    assertTrue(
                        typed.stream().map(SchemaValidation::pointer).anyMatch(pointer::equals),
                        () ->
                            "expected exact pointer "
                                + pointer
                                + " in "
                                + SchemaValidation.describe(typed));
                  }
                }
                if (reject) {
                  ProtocolException ex =
                      assertThrows(ProtocolException.class, () -> ProtocolJson.read(text));
                  assertEquals(reason, ex.reason().name());
                  if (pointer != null && ex.pointer() != null) {
                    assertTrue(
                        ex.pointer().startsWith(pointer),
                        () -> "codec pointer " + ex.pointer() + " must refine " + pointer);
                  }
                } else {
                  ProtocolJson.read(text);
                }
              }));
    }
    return tests;
  }

  @TestFactory
  List<DynamicTest> runs() {
    List<DynamicTest> tests = new ArrayList<>();
    for (JsonNode run : Corpus.manifest().get("runs")) {
      String dir = run.get("dir").asText();
      boolean invalid = "INVALID".equals(run.get("outcome").asText());
      String reason = run.hasNonNull("reason") ? run.get("reason").asText() : null;
      int badLine = run.hasNonNull("line") ? run.get("line").asInt() : -1;
      boolean idempotentOnly = run.hasNonNull("roundTrip");
      tests.add(
          dynamicTest(
              dir,
              () -> {
                Path runDir = Corpus.fixtures().resolve(dir);
                List<String> lines = Corpus.runLines(runDir);
                for (int i = 0; i < lines.size(); i++) {
                  String line = lines.get(i);
                  int number = i + 1;
                  if (invalid
                      && reason != null
                      && CODEC_REASONS.contains(reason)
                      && number == badLine) {
                    ProtocolException ex =
                        assertThrows(ProtocolException.class, () -> ProtocolJson.read(line));
                    assertEquals(reason, ex.reason().name());
                    continue;
                  }
                  Event event = ProtocolJson.read(line);
                  if (!(event.payload() instanceof UnknownPayload)) {
                    assertTrue(
                        SchemaValidation.isValid(line),
                        () ->
                            dir
                                + ":"
                                + number
                                + " "
                                + SchemaValidation.describe(SchemaValidation.errors(line)));
                  }
                  String written = ProtocolJson.write(event);
                  if (!idempotentOnly) {
                    assertEquals(
                        Corpus.canonical(line), Corpus.canonical(written), dir + ":" + number);
                  }
                }
              }));
    }
    return tests;
  }

  @TestFactory
  List<DynamicTest> limitsBeyondTheCommittedCorpus() {
    String base =
        Corpus.readText(
            Corpus.fixtures().resolve("events/valid/attempt-finished-failed-full.json"));
    Event event = ProtocolJson.read(base);
    AttemptFinished payload = assertInstanceOf(AttemptFinished.class, event.payload());
    return List.of(
        dynamicTest(
            "message at 65536 accepted, 65537 rejected",
            () -> {
              assertTrue(SchemaValidation.isValid(withMessage(event, payload, "m".repeat(65536))));
              assertFalse(SchemaValidation.isValid(withMessage(event, payload, "m".repeat(65537))));
            }),
        dynamicTest(
            "stack trace at 262144 accepted, 262145 rejected",
            () -> {
              assertTrue(SchemaValidation.isValid(withStack(event, payload, "s".repeat(262144))));
              assertFalse(SchemaValidation.isValid(withStack(event, payload, "s".repeat(262145))));
            }));
  }

  private static String withMessage(Event e, AttemptFinished p, String message) {
    return ProtocolJson.write(replacePayload(e, p, List.of(Failure.of(message))));
  }

  private static String withStack(Event e, AttemptFinished p, String stack) {
    return ProtocolJson.write(
        replacePayload(e, p, List.of(new Failure("x", null, stack, null, null))));
  }

  private static Event replacePayload(Event e, AttemptFinished p, List<Failure> failures) {
    AttemptFinished np =
        new AttemptFinished(
            p.attemptId(), p.status(), p.rawStatus(), p.expectedStatus(), p.durationMs(), failures);
    return new Event(
        e.protocolVersion(),
        e.eventId(),
        e.eventType(),
        e.runId(),
        e.sessionId(),
        e.sequence(),
        e.occurredAt(),
        e.ignorable(),
        np);
  }
}

File: ts/packages/protocol/test/corpus.test.ts
import { Ajv2020 } from 'ajv/dist/2020.js';
import addFormats from 'ajv-formats';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { describe, expect, it } from 'vitest';
import {
  EVENT_TYPES,
  ProtocolError,
  isKnownEvent,
  isKnownEventType,
  parseEvent,
  stringifyEvent,
} from '../src/index.js';
import { PROTOCOL_DIR, canonical, fixtureText, manifest, runLines } from './helpers.js';

const schema = JSON.parse(
  readFileSync(join(PROTOCOL_DIR, 'schema', 'event.schema.json'), 'utf8'),
) as object;
const ajv = new Ajv2020({ allErrors: true });
addFormats.default(ajv);
const validate = ajv.compile(schema);
const { $id: _id, oneOf: _oneOf, ...base } = schema as Record<string, unknown>;
void _id;
void _oneOf;
/** The branch for one event type yields exact pointers instead of the oneOf's root error. */
const byType = new Map(
  EVENT_TYPES.map((t) => [t, ajv.compile({ ...base, allOf: [{ $ref: `#/$defs/event.${t}` }] })]),
);
const m = manifest();
const CODEC_REASONS = new Set([
  'MALFORMED_JSON',
  'SCHEMA_INVALID',
  'UNSUPPORTED_PROTOCOL_VERSION',
  'UNSUPPORTED_EVENT_TYPE',
]);

describe('valid event fixtures', () => {
  for (const { file, roundTrip } of m.events.valid) {
    it(`${file} validates and round-trips (${roundTrip})`, () => {
      const text = fixtureText(file);
      const original = JSON.parse(text) as { ignorable?: boolean };
      const event = parseEvent(text);
      if (isKnownEvent(event))
        expect(validate(original), JSON.stringify(validate.errors)).toBe(true);
      else expect(original.ignorable).toBe(true);
      const written = stringifyEvent(event);
      if (isKnownEvent(event)) expect(validate(JSON.parse(written))).toBe(true);
      if (roundTrip === 'exact') expect(canonical(JSON.parse(written))).toBe(canonical(original));
      expect(canonical(JSON.parse(stringifyEvent(parseEvent(written))))).toBe(
        canonical(JSON.parse(written)),
      );
    });
  }
});

describe('invalid event fixtures', () => {
  for (const { file, reason, pointer, codec } of m.events.invalid) {
    it(`${file} is rejected for ${reason}`, () => {
      const text = fixtureText(file);
      if (reason === 'SCHEMA_INVALID') {
        const parsed = JSON.parse(text) as { eventType?: string };
        expect(validate(parsed)).toBe(false);
        const typed =
          parsed.eventType !== undefined && isKnownEventType(parsed.eventType)
            ? byType.get(parsed.eventType)
            : undefined;
        const v = typed ?? validate;
        expect(v(parsed)).toBe(false);
        if (pointer !== null) {
          const paths = (v.errors ?? []).map((e) => e.instancePath);
          expect(paths, paths.join(',')).toContain(pointer);
        }
      }
      if (codec === 'reject') {
        expect(() => parseEvent(text)).toThrow(ProtocolError);
        try {
          parseEvent(text);
        } catch (e) {
          expect((e as ProtocolError).code).toBe(reason);
          const codecPointer = (e as ProtocolError).pointer;
          if (pointer !== null && codecPointer !== undefined)
            expect(
              codecPointer.startsWith(pointer),
              `codec pointer ${codecPointer} must refine ${pointer}`,
            ).toBe(true);
        }
      } else {
        expect(() => parseEvent(text)).not.toThrow();
      }
    });
  }
});

describe('run fixtures parse line by line', () => {
  for (const run of m.runs) {
    it(run.dir, () => {
      const lines = runLines(run.dir);
      lines.forEach((line, i) => {
        const number = i + 1;
        if (
          run.outcome === 'INVALID' &&
          run.reason !== undefined &&
          CODEC_REASONS.has(run.reason) &&
          run.line === number
        ) {
          expect(() => parseEvent(line)).toThrow(ProtocolError);
          try {
            parseEvent(line);
          } catch (e) {
            expect((e as ProtocolError).code).toBe(run.reason);
          }
          return;
        }
        if (
          run.outcome === 'INVALID' &&
          run.reason === 'MALFORMED_JSON' &&
          run.line !== number &&
          (() => {
            try {
              JSON.parse(line);
              return false;
            } catch {
              return true;
            }
          })()
        )
          return;
        const event = parseEvent(line);
        if (isKnownEvent(event))
          expect(
            validate(JSON.parse(line)),
            `${run.dir}:${number} ${JSON.stringify(validate.errors)}`,
          ).toBe(true);
        const written = stringifyEvent(event);
        if (run.roundTrip !== 'idempotent')
          expect(canonical(JSON.parse(written))).toBe(canonical(JSON.parse(line)));
      });
    });
  }
});

describe('limits beyond the committed corpus', () => {
  const base = JSON.parse(fixtureText('events/valid/attempt-finished-failed-full.json')) as {
    payload: { failures: { message: string; stackTrace?: string }[] };
  };
  it('accepts a failure message at 65536 characters and rejects 65537', () => {
    const at = structuredClone(base);
    at.payload.failures = [{ message: 'm'.repeat(65536) }];
    expect(validate(at)).toBe(true);
    at.payload.failures = [{ message: 'm'.repeat(65537) }];
    expect(validate(at)).toBe(false);
  });
  it('accepts a stack trace at 262144 characters and rejects 262145', () => {
    const at = structuredClone(base);
    at.payload.failures = [{ message: 'x', stackTrace: 's'.repeat(262144) }];
    expect(validate(at)).toBe(true);
    at.payload.failures = [{ message: 'x', stackTrace: 's'.repeat(262145) }];
    expect(validate(at)).toBe(false);
  });
});

describe('string maps keep every key as data', () => {
  const mapText = '{"__proto__":"one","constructor":"two","prototype":"three","normal":"four"}';
  // Written by hand: an object literal with a `__proto__:` entry would set the prototype instead.
  const wire =
    '{"protocolVersion":"0.3.0","eventId":"e-1","eventType":"session.started","runId":"r",' +
    '"sessionId":"s","sequence":1,"occurredAt":"2026-01-01T00:00:00Z","payload":{"producer":' +
    `{"name":"p"},"environment":${mapText},"labels":${mapText}}}`;

  it('decodes __proto__ as an own enumerable string property without touching the prototype', () => {
    const before = Object.getOwnPropertyNames(Object.prototype).sort();
    const event = parseEvent(wire);
    if (event.eventType !== 'session.started') throw new Error('unexpected type');
    for (const m of [event.payload.environment, event.payload.labels]) {
      const map = m as Record<string, string>;
      expect(Object.getPrototypeOf(map)).toBe(Object.prototype);
      expect(Object.prototype.hasOwnProperty.call(map, '__proto__')).toBe(true);
      expect(Object.getOwnPropertyDescriptor(map, '__proto__')).toMatchObject({
        value: 'one',
        enumerable: true,
        writable: true,
        configurable: true,
      });
      expect(Object.keys(map)).toEqual(['__proto__', 'constructor', 'prototype', 'normal']);
      expect([map['__proto__'], map['constructor'], map['prototype'], map['normal']]).toEqual([
        'one',
        'two',
        'three',
        'four',
      ]);
      expect(JSON.stringify(map)).toBe(mapText);
    }
    expect(Object.getOwnPropertyNames(Object.prototype).sort()).toEqual(before);
    expect(Object.getPrototypeOf({})).toBe(Object.prototype);
  });

  it('round-trips such maps through the codec unchanged', () => {
    const written = stringifyEvent(parseEvent(wire));
    expect(written).toContain('"__proto__":"one"');
    expect(canonical(JSON.parse(written))).toBe(canonical(JSON.parse(wire)));
    expect(canonical(JSON.parse(stringifyEvent(parseEvent(written))))).toBe(
      canonical(JSON.parse(wire)),
    );
  });
});
